### PR TITLE
Add Logging and PubSub documentation

### PR DESCRIPTION
**Pull request by quoctruong**
image1
image2
image3
image4

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/google-cloud-powershell/446)

### CODE REVIEW

File: js/app/controllers/content-product.ng [hunk 1]
```diff
@@ -209,6 +209,158 @@ Import-GcSqlInstance "my-instance-name" "C:\Users\User\file.csv" `
...
 
+  <!-- Google Cloud PubSub -->
+  <div ng-switch-when="google-cloud-pubsub">
+    <p>
+      Google Cloud Pub/Sub is a fully-managed real-time messaging service that
+      allows you to send and receive messages between independent applications.
+    </p>
+    <h2>Publisher</h2>
+    <p>
+      Publisher application creates and sends messages to a <em>topic</em>.
```
Comment (iantalarico) [inline js/app/controllers/content-product.ng]: The publisher application? or Publisher applications creat and send?

File: js/app/controllers/content-product.ng [hunk 2]
```diff
@@ -209,6 +209,158 @@ Import-GcSqlInstance "my-instance-name" "C:\Users\User\file.csv" `
...
+      The <cmdlet-ref name="New-GcpsTopic"></cmdlet-ref> can be called to create an instance
+      in a particular topic. If your active gcloud configuration has a project, you don't have to
+      use <code>-Project</code> parameter.
```
Comment (iantalarico) [inline js/app/controllers/content-product.ng]: use the?

use parameter <code>-Project</code>?

File: js/app/controllers/content-product.ng [hunk 3]
```diff
@@ -209,6 +209,158 @@ Import-GcSqlInstance "my-instance-name" "C:\Users\User\file.csv" `
...
+    </p>
+    <pre>
+# Creates topic "my-topic" in the default project.
+New-GcpsTopic -Topic "my-topic"</pre>
+    <p>
+      After the topic has been created, you can now publish messages to the topic using the
+      <cmdlet-ref name="Publish-GcpsMessage"></cmdlet-ref> cmdlet.
+    </p>
+    <pre>
+# Publish message with data "This is a test" to topic "my-topic".
```
Comment (iantalarico) [inline js/app/controllers/content-product.ng]: Publishes the message?

File: js/app/controllers/content-product.ng [hunk 4]
```diff
@@ -209,6 +209,158 @@ Import-GcSqlInstance "my-instance-name" "C:\Users\User\file.csv" `
...
+    <pre>
+# Creates two messages.
+$messageOne = New-GcpsMessage -Data "This is a test"
+$messageTwo = New-GcpsMessage -Data "Data" -Attributes @{"key" = "value"}
+      
+# Publish the messages to topic "my-topic".
+Publish-GcpsMessage -Message @($messageOne, $messageTwo) -Topic "my-topic"</pre>
+    <h2>Subscriber</h2>
+    <p>
+      Subscriber application creates a <em>subscription</em> to a topic to receive messages from it.
```
Comment (iantalarico) [inline js/app/controllers/content-product.ng]: The subscriber application? or Subscriber applications create?

File: js/app/controllers/content-product.ng [hunk 5]
```diff
@@ -209,6 +209,158 @@ Import-GcSqlInstance "my-instance-name" "C:\Users\User\file.csv" `
...
+      The <cmdlet-ref name="New-GcpsSubscription"></cmdlet-ref> cmdlet can be called to create an instance
+      in a particular topic. If your active gcloud configuration has a project, you don't have to
+      use <code>-Project</code> parameter.
```
Comment (iantalarico) [inline js/app/controllers/content-product.ng]: use the?

use parameter -Project?

File: js/app/controllers/content-product.ng [hunk 6]
```diff
@@ -209,6 +209,158 @@ Import-GcSqlInstance "my-instance-name" "C:\Users\User\file.csv" `
...
+    </p>
+    <pre>
+# Creates pull subscription "pull-subscription" to topic "my-topic" in the default project.
+New-GcpsSubscription -Topic "my-topic" -Subscription "pull-subscription"
+
+# Creates push subscription "push-subscription" to topic "my-topic".
+New-GcpsSubscription -Topic "my-topic" `
+                     -Subscription "push-subscription" `
+                     -PushEndpoint "http://www.example.com"
+</pre>
```
Comment (iantalarico) [inline js/app/controllers/content-product.ng]: nit: indentation is off

File: js/app/controllers/content-product.ng [hunk 7]
```diff
@@ -209,6 +209,158 @@ Import-GcSqlInstance "my-instance-name" "C:\Users\User\file.csv" `
...
+      To pull messages from a subscription, the <cmdlet-ref name="Get-GcpsMessage"></cmdlet-ref> cmdlet can
+      be used. By default, the cmdlet will block until at least one message is retrieved. To prevent blocking,
+      the switch <code>-ReturnImmediately</code> can be used. The cmdlet can also automatically send
+      an acknowledgement for every retrieved message if the switch <code>-AutoAck</code> is used. If not,
+      you will have to use the <cmdlet-ref name="Send-GcpsAck"></cmdlet-ref> cmdlet to send the acknowledgement.
+      Unacknowledged messages will become available again for pulling after the acknowledgement deadline of the message expires.
+    </p>
+    <pre>
+# Pulls messages from subscription "my-subscription" and sends out acknowledgement automatically.
+Get-GcpsMessage -Subscription "--subscription" -AutoAck
```
Comment (iantalarico) [inline js/app/controllers/content-product.ng]: Should "my-subscription" be included here?

File: js/app/controllers/content-product.ng [hunk 8]
```diff
@@ -209,6 +209,158 @@ Import-GcSqlInstance "my-instance-name" "C:\Users\User\file.csv" `
...
+
+# Pulls messages from subscription "my-subscription" and sends out acknowledgement with Send-GcpsAck.
+$messages = Get-GcpsMessage -Subscription "--subscription"
```
Comment (iantalarico) [inline js/app/controllers/content-product.ng]: Should "my-subscription" be included here?

File: js/app/controllers/content-product.ng [hunk 9]
```diff
@@ -209,6 +209,158 @@ Import-GcSqlInstance "my-instance-name" "C:\Users\User\file.csv" `
...
+Send-GcpsAck -InputObject $messages</pre>
```
Comment (iantalarico) [inline js/app/controllers/content-product.ng]: nit: pre is on same line

**Comment by quoctruong** [comment]
Thanks!

File: js/app/controllers/content-product.ng [hunk 10]
```diff
@@ -209,6 +209,158 @@ Import-GcSqlInstance "my-instance-name" "C:\Users\User\file.csv" `
...
+  </div>
+
+  <!-- Google Cloud Logging -->
+  <div ng-switch-when="google-cloud-logging">
+    <p>
+      Stackdriver Logging allows you to store, search, analyze, monitor and alert on log data
+      and events from Google Cloud Platform and Amazon Web Services.
+    </p>
+    <h2>Log and log entries</h2>
```
Comment (iantalarico) [inline js/app/controllers/content-product.ng]: Logs?

Should this be Logs and Log Entries? All caps for title?

File: js/app/controllers/content-product.ng [hunk 11]
```diff
@@ -209,6 +209,158 @@ Import-GcSqlInstance "my-instance-name" "C:\Users\User\file.csv" `
...
+      You will have to specify the log that the entry belongs to (if the log does not exist, it will
+      be created). To associate the log with a monitored resource, you can use the -MonitoredResource parameter.
+      By default, the log entry is associated with the "global" resource. To create a monitored resource,
+      use the <cmdlet-ref name="New-GcLogMonitoredResource"></cmdlet-ref> cmdlet.
+  </p>
+    <pre>
+# Creates a log entry in the log "my-log".
+New-GcLogEntry -LogName "my-log" -TextPayload "This is a log."
+
+# Creates a log entry associated with Cloud SQL monitored resource.
```
Comment (iantalarico) [inline js/app/controllers/content-product.ng]: with a

File: js/app/controllers/content-product.ng [hunk 12]
```diff
@@ -209,6 +209,158 @@ Import-GcSqlInstance "my-instance-name" "C:\Users\User\file.csv" `
...
+$resource = New-GcLogMonitoredResource -ResourceType "cloudsql_database" `
+                                       -Labels @{"project_id" = "my-project";
+                                                 "database_id" = "id"}
+New-GcLogEntry -LogName "my-log" `
+               -TextPayload "Log" `
```
Comment (iantalarico) [inline js/app/controllers/content-product.ng]: I would change this payload to be a little larger to not be confusing what needs to be here

File: js/app/controllers/content-product.ng [hunk 13]
```diff
@@ -209,6 +209,158 @@ Import-GcSqlInstance "my-instance-name" "C:\Users\User\file.csv" `
...
+               -MonitoredResource $resource</pre>
```
Comment (iantalarico) [inline js/app/controllers/content-product.ng]: nit: pre is on same line

File: js/app/controllers/content-product.ng [hunk 14]
```diff
@@ -209,6 +209,158 @@ Import-GcSqlInstance "my-instance-name" "C:\Users\User\file.csv" `
...
+    <p>
+      You can retrieve log entries with the cmdlet <cmdlet-ref name="Get-GcLogEntry"></cmdlet-ref>.
+    </p>
+    <pre>
+# Gets all entries from log "my-log"
+Get-GcLogEntry -LogName "my-log"
+
+# Gets all entries associated with Google Cloud Engine instances.
+Get-GcLogEntry -ResourceName "gce_instance"</pre>
```
Comment (iantalarico) [inline js/app/controllers/content-product.ng]: nit: pre is on same line

File: js/app/controllers/content-product.ng [hunk 15]
```diff
@@ -209,6 +209,158 @@ Import-GcSqlInstance "my-instance-name" "C:\Users\User\file.csv" `
...
+      Stackdriver Logging will match incoming log entries against your sinks and all log entries matching each sink
+      are then copied to the associated destination. Log entries that exist before the sink is created will not be exported.
+    </p>
+    <p>
+      Destinations for exported logs can be Google Cloud Storage Buckets, Google BigQuery Datasets
+      or Google Cloud Pub/Sub Topics.
+    </p>
+    <pre>
+# Creates a log sink for log entries in the default project.
+New-GcLogSink -Sink "my-sink" -GcsBucketDestination "my-bucket"
```
Comment (iantalarico) [inline js/app/controllers/content-product.ng]: maybe add to be sent to google cloud storage?

File: js/app/controllers/content-product.ng [hunk 16]
```diff
@@ -209,6 +209,158 @@ Import-GcSqlInstance "my-instance-name" "C:\Users\User\file.csv" `
...
+
+# Creates a log sink for log entries in log "my-log".
```
Comment (iantalarico) [inline js/app/controllers/content-product.ng]: maybe add to be sent to biq query?

File: js/app/controllers/content-product.ng [hunk 17]
```diff
@@ -209,6 +209,158 @@ Import-GcSqlInstance "my-instance-name" "C:\Users\User\file.csv" `
...
+New-GcLogSink -Sink "my-sink" `
+              -LogName "my-log" `
+              -BigQueryDataSetDestination "my_dataset"
+
+# Creates a log sink for log entries that match the filter.
+New-GcLogSink -Sink "my-sink" `
```
Comment (iantalarico) [inline js/app/controllers/content-product.ng]: maybe add to be sent to pubsub?

**Comment by quoctruong** [comment]
@iantalarico Sorry I closed the PR before you commented. I opened another PR to address the comment here https://github.com/GoogleCloudPlatform/google-cloud-powershell/pull/448.

**Pull request merged**